File: scripts/coverage_t_test.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys

import pandas as pd
import pysam
from scipy.stats import fisher_exact


def coverage_t_test(bam, taxon_fasta_idx, taxon, outfile):
    """
    Function that calculates the pvalue of the coverage. Testing if there was clustering bias during the
    sequencing/the reads that contribute to the identification/abundance of a species come only
    from a specific genomic region
    """
    taxon_seqlen = genome_sizes(taxon_fasta_idx)

    pileup_dict = {}

    bamfile = pysam.AlignmentFile(bam, "rb")

    for pileupcolumn in bamfile.pileup():
        pileup_dict[pileupcolumn.pos] = pileupcolumn.n

    expected_coverage = sum(list(pileup_dict.values()))
    observed_coverage = len(list(pileup_dict.keys()))

    contingency_first_row = [observed_coverage, expected_coverage]
    print("Observed and expected coverage are ", contingency_first_row, file=sys.stderr)

    contingency_second_row = [taxon_seqlen, taxon_seqlen]

    oddsratio, pvalue = fisher_exact([contingency_first_row, contingency_second_row])

    with open(outfile, 'w') as outhandle:
        print(taxon, pvalue, file=outhandle, sep='\t')


def genome_sizes(taxon_fasta_idx):
    faidx = pd.read_csv(taxon_fasta_idx, sep='\t', names=['Name', 'Length', 'Offset', 'Linebases', 'Linewidth'])

    taxon_seq_len = faidx['Length'].sum()

    return taxon_seq_len


if __name__ == '__main__':
    # redirect all output to the log
    sys.stderr = open(snakemake.log[0], 'w')

    coverage_t_test(
        bam=snakemake.input[0],
        taxon_fasta_idx=snakemake.input[1],
        taxon=snakemake.wildcards.orgname,
        outfile=snakemake.output[0]
    )

File: haystack/workflow/scripts/coverage_t_test.R
# Title     : coverage_t_test
# Objective : Function that calculates the pvalue of the coverage. Testing if there was clustering bias during the
#     sequencing/the reads that contribute to the identification/abundance of a species come only
#     from a specific genomic region
# Created by: Evangelos A. Dimopoulos, Evan K. Irving-Pease
# Created on: 10/11/2020

cov_file <- snakemake@input[[1]]
taxon_fasta_idx <- snakemake@input[[2]]
taxon <- snakemake@wildcards[["orgname"]]
outfile <- snakemake@output[[1]]

file_empty <- function(filenames) file.info(filenames)$size == 0

genome_sizes <- function(taxon_fasta_idx) {
  faidx <- read.csv(taxon_fasta_idx, sep='\t')
  names(faidx) <- c('Name','Length','Offset', 'Linebases', 'Linewidth')
  taxon_seq_len <- sum(faidx[,"Length"])
  return(taxon_seq_len)
}

# check if the coverage stats file is empty

if (file_empty(cov_file)) {
  stop(paste("The file with the coverage stats", cov_file, "is empty.", sep = " "))
}

# get the total length of the ref genome
taxon_seqlen <- genome_sizes(taxon_fasta_idx)

# read the cov stats file
cov_stats <- read.csv(cov_file, sep='\t')
names(cov_stats) <- c('observed','expected')

expected_coverage <- cov_stats[1,"expected"]
observed_coverage <- cov_stats[1,"observed"]

contingency_first_row <- c(observed_coverage, expected_coverage)

print(paste("Observed and expected coverage are", contingency_first_row, sep='\t'))

contingency_second_row <- c(taxon_seqlen, taxon_seqlen)

test_input <- matrix(rbind(contingency_first_row, contingency_second_row), nrow=2)

#perform the test
pvalue <- fisher.test(test_input)$p.value

#write to the output file
write.table(data.frame(taxon, pvalue), file=outfile,
            sep = "\t", row.names=FALSE, col.names=FALSE, quote=FALSE)

